server/games/game_results.py

```diff
@@ -0,0 +1,136 @@
+from enum import Enum
+from collections import Counter
+from collections.abc import Mapping
+from server.decorators import with_logger
+from typing import NamedTuple
+
+
+class GameOutcome(Enum):
+    VICTORY = 'victory'
+    DEFEAT = 'defeat'
+    DRAW = 'draw'
+    MUTUAL_DRAW = 'mutual_draw'
+    UNKNOWN = 'unknown'
+
+    @classmethod
+    def from_message(cls, msg: str):
+        try:
+            return cls(msg.lower())
+        except ValueError:
+            return cls.UNKNOWN
+
+
+class GameResult(NamedTuple):
+    """
+    These are sent from each player's FA when they quit the game. 'Score'
+    depends on the number of ACUs killed, whether the player died, maybe other
+    factors.
+    """
+    reporter: int
+    army: int
+    outcome: GameOutcome
+    score: int
+
+
+@with_logger
+class GameResults(Mapping):
+    """
+    Collects all results from a single game. Allows to determine results for an
+    army and game as a whole. Supports a dict-like access to lists of results
+    for each army, but don't modify these.
+    """
+    def __init__(self):
+        Mapping.__init__(self)
+        self._back = {}
+
+    def __getitem__(self, key: int):
+        return self._back[key]
+
+    def __iter__(self):
+        return iter(self._back)
+
+    def __len__(self):
+        return len(self._back)
+
+    def add(self, result: GameResult):
+        army_results = self._back.setdefault(result.army, [])
+        army_results.append(result)
+
+    def is_mutually_agreed_draw(self, player_armies):
+        # Can't tell if we have no results
+        if not self:
+            return False
+        # Everyone has to agree to a mutual draw
+        for army in player_armies:
+            if army not in self:
+                continue
+            if any(r.outcome is not GameOutcome.MUTUAL_DRAW
+                   for r in self[army]):
+                return False
+        return True
+
+    def outcome(self, army: int) -> GameOutcome:
+        """
+        Determines what the game outcome was for a given army. Returns the
+        outcome all players agree on, excluding players that reported an
+        unknown outcome. Reports unknown outcome if players disagree.
+        """
+        if army not in self:
+            return GameOutcome.UNKNOWN
+
+        outcomes = set(r.outcome for r in self[army]
+                       if r.outcome is not GameOutcome.UNKNOWN)
+
+        if len(outcomes) == 1:
+            return outcomes.pop()
+        else:
```

Comment:
Can drop the else

Reply:
The else clause is a one-liner, so I thought it's clearer to keep it.

Reply:
Now that it's not a one liner you should probably get rid of the else. It's just an unnecessary level of indentation